homeassistant/components/monoprice/media_player.py

```diff
@@ -36,104 +29,89 @@
     | SUPPORT_SELECT_SOURCE
 )
 
-ZONE_SCHEMA = vol.Schema({vol.Required(CONF_NAME): cv.string})
-
-SOURCE_SCHEMA = vol.Schema({vol.Required(CONF_NAME): cv.string})
 
-CONF_ZONES = "zones"
-CONF_SOURCES = "sources"
+def _get_sources(sources_config):
+    source_id_name = {int(index): name for index, name in sources_config.items()}
 
-DATA_MONOPRICE = "monoprice"
+    source_name_id = {v: k for k, v in source_id_name.items()}
 
-# Valid zone ids: 11-16 or 21-26 or 31-36
-ZONE_IDS = vol.All(
-    vol.Coerce(int),
-    vol.Any(
-        vol.Range(min=11, max=16), vol.Range(min=21, max=26), vol.Range(min=31, max=36)
-    ),
-)
+    source_names = sorted(source_name_id.keys(), key=lambda v: source_name_id[v])
 
-# Valid source ids: 1-6
-SOURCE_IDS = vol.All(vol.Coerce(int), vol.Range(min=1, max=6))
+    return [source_id_name, source_name_id, source_names]
 
-MEDIA_PLAYER_SCHEMA = vol.Schema({ATTR_ENTITY_ID: cv.comp_entity_ids})
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Required(CONF_PORT): cv.string,
-        vol.Required(CONF_ZONES): vol.Schema({ZONE_IDS: ZONE_SCHEMA}),
-        vol.Required(CONF_SOURCES): vol.Schema({SOURCE_IDS: SOURCE_SCHEMA}),
-    }
-)
 
-
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_entry(hass, config_entry, async_add_devices):
     """Set up the Monoprice 6-zone amplifier platform."""
-    port = config.get(CONF_PORT)
+    port = config_entry.data.get(CONF_PORT)
 
     try:
-        monoprice = get_monoprice(port)
+        monoprice = await hass.async_add_executor_job(get_monoprice, port)
```

Comment:
If we do this in `__init__.py` we can `raise ConfigEntryNotReady` so it will automatically retry later.

Reply:
Maybe use `async_get_monoprice` instead?

Reply:
@MartinHjelmare that's my goal, and the code changes are relatively straightforward. The reason I'm not doing this yet is that `pymonoprice` is dependent on `pyserial-asyncio` for its async implementation, which does not seem to be maintained and is using old-style coroutine syntax. I'm afraid it's not going to work with future python versions.